### PR TITLE
 👷 Add client document

### PR DESCRIPTION
**Pull request by kiyonlin**

### CODE REVIEW

File: api/client.md [hunk 1]
```diff
@@ -0,0 +1,646 @@
+---
+description: >-
+  The Client struct represents the HTTP Client. It has methods like Get, Post, etc, they'll create a new Agent instance and parse request url automatically.
```
Comment (hi019) [inline api/client.md:3]:
```suggestion
  The Client struct represents the Fiber HTTP Client.
  
  
```

File: api/client.md [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+func (c *Client) Head(url string) *Agent
+func (c *Client) Post(url string) *Agent
+func (c *Client) Put(url string) *Agent
+func (c *Client) Patch(url string) *Agent
+func (c *Client) Delete(url string) *Agent
+```
+{% endcode %}
+
+## ✨ Agent
+`Agent` is on top of [fasthttp `HostClient`](https://github.com/valyala/fasthttp/blob/master/client.go#L603) which has lots of convenient helper methods and several `send request` methods.
```
Comment (hi019) [inline api/client.md:24]:
```suggestion
`Agent` is built on top of FastHTTP's [`HostClient`](https://github.com/valyala/fasthttp/blob/master/client.go#L603) which has lots of convenient helper methods and several `send request` methods.
```

Comment (hi019) [inline api/client.md:24]: By send request methods do you mean like GET & POST?

File: api/client.md [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+
+### Parse
+Parse initializes URI and HostClient.
```
Comment (hi019) [inline api/client.md:27]: Sounds like Parse takes a URI as a param

```suggestion
Parse initializes a HostClient.
```

File: api/client.md [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+if err := a.Parse(); err != nil {
+	panic(err)
+}
+
+code, body, errs := a.Bytes() // ...
+```
+{% endcode %}
+
+### Set
+Set sets the given 'key: value' header.
```
Comment (hi019) [inline api/client.md:45]:
```suggestion
Set sets the given `key: value` header.
```

File: api/client.md [hunk 5]
```diff
@@ -0,0 +1,646 @@
...
+agent.Set("k1", "v1").
+    SetBytesK([]byte("k1"), "v1").
+    SetBytesV("k1", []byte("v1")).
+    SetBytesKV([]byte("k2"), []byte("v2"))
+// ...
+```
+{% endcode %}
+
+### Add
+Add adds the given 'key: value' header. Multiple headers with the same key may be added with this function.
```
Comment (hi019) [inline api/client.md:68]:
```suggestion
Add adds the given `key: value` header. Multiple headers with the same key may be added with this function.
```

Comment (hi019) [inline api/client.md:68]: > Multiple headers with the same key may be added with this function.
So does this mean the header's won't overwrite eachother?

File: api/client.md [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+{% endcode %}
+
+{% code title="Example" %}
+```go
+
+agent.Add("k1", "v1").
+    AddBytesK([]byte("k1"), "v1").
+    AddBytesV("k1", []byte("v1")).
+    AddBytesKV([]byte("k2"), []byte("v2"))
+// ...
```
Comment (hi019) [inline api/client.md:86]: We should show what the headers look like:


```suggestion
// Headers:
// k1: v1
// k1: v1
// k2: v2
```

File: api/client.md [hunk 7]
```diff
@@ -0,0 +1,646 @@
...
+```
+{% endcode %}
+
+### ConnectionClose
+ConnectionClose sets 'Connection: close' header.
```
Comment (hi019) [inline api/client.md:91]:
```suggestion
ConnectionClose adds the `Connection: close` header.
```

File: api/client.md [hunk 8]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.ConnectionClose()
+// ...
+```
+{% endcode %}
+
+### UserAgent
+UserAgent sets User-Agent header value.
```
Comment (hi019) [inline api/client.md:107]:
```suggestion
UserAgent sets `User-Agent` header value.
```

File: api/client.md [hunk 9]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.UserAgent("fiber")
+// ...
+```
+{% endcode %}
+
+### Cookie
+Cookie sets 'key: value' cookie.
```
Comment (hi019) [inline api/client.md:124]:
```suggestion
Cookie sets a cookie in `key: value` form. `Cookies` can be used to set multiple cookies.
```

File: api/client.md [hunk 10]
```diff
@@ -0,0 +1,646 @@
...
+func (a *Agent) CookieBytesKV(key, value []byte) *Agent
+func (a *Agent) Cookies(kv ...string) *Agent
+func (a *Agent) CookiesBytesKV(kv ...[]byte) *Agent
+```
+{% endcode %}
+
+{% code title="Example" %}
+```go
+agent.Cookie("k", "v")
+agent.Cookies("k1", "v1", "k2", "v2")
```
Comment (hi019) [inline api/client.md:139]: Are we able use a map too?

```
m := map[string]string{
  "k1": "v1",
  "k2": "v2",
}
agent.Cookies(m)
```

File: api/client.md [hunk 11]
```diff
@@ -0,0 +1,646 @@
...
+// ...
+```
+{% endcode %}
+
+### Referer
+Referer sets Referer header value.
```
Comment (hi019) [inline api/client.md:145]:
```suggestion
Referer sets the Referer header value.
```

File: api/client.md [hunk 12]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.ContentType("custom-type")
+// ...
+```
+{% endcode %}
+
+### Host
+Host sets host for the uri.
```
Comment (hi019) [inline api/client.md:179]:
```suggestion
Host sets the Host header.
```

File: api/client.md [hunk 13]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.Host("example.com")
+// ...
+```
+{% endcode %}
+
+### QueryString
+QueryString sets URI query string.
```
Comment (hi019) [inline api/client.md:196]:
```suggestion
QueryString sets the URI query string.
```

File: api/client.md [hunk 14]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.QueryString("foo=bar")
+// ...
+```
+{% endcode %}
+
+### BasicAuth
+BasicAuth sets URI username and password.
```
Comment (hi019) [inline api/client.md:213]:
```suggestion
BasicAuth sets the URI username and password using HTTP Basic Auth.
```

File: api/client.md [hunk 15]
```diff
@@ -0,0 +1,646 @@
...
+```go
+agent.BodyString("foo=bar")
+agent.Body([]byte("bar=baz"))
+agent.BodyStream(strings.NewReader("body=stream"), -1)
+// ...
+```
+{% endcode %}
+
+### JSON
+JSON sends a JSON request with `application/json` content type.
```
Comment (hi019) [inline api/client.md:262]:
```suggestion
JSON sends a JSON request by setting the Content-Type header to `application/json`.
```

File: api/client.md [hunk 16]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.JSON(fiber.Map{"success": true})
+// ...
+```
+{% endcode %}
+
+### XML
+XML sends an XML request with `application/xml` content type.
```
Comment (hi019) [inline api/client.md:278]:
```suggestion
XML sends an XML request by setting the Content-Type header to `application/xml`.
```

File: api/client.md [hunk 17]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Example" %}
+```go
+agent.XML(fiber.Map{"success": true})
+// ...
+```
+{% endcode %}
+
+### Form
+Form sends form request with `application/x-www-form-urlencoded` content type.
```
Comment (hi019) [inline api/client.md:294]:
```suggestion
Form sends a form request by setting the Content-Type header to `application/x-www-form-urlencoded`.
```

File: api/client.md [hunk 18]
```diff
@@ -0,0 +1,646 @@
...
+args.Set("foo", "bar")
+
+agent.Form(args)
+// ...
+ReleaseArgs(args)
+```
+{% endcode %}
+
+### MultipartForm
+MultipartForm sends multipart form request with `multipart/form-data` content type, k-v and files.
```
Comment (hi019) [inline api/client.md:318]:
```suggestion
MultipartForm sends multipart form request by setting the Content-Type header to `multipart/form-data`. These requests can include key-value's and files.
```

File: api/client.md [hunk 19]
```diff
@@ -0,0 +1,646 @@
...
+
+{% code title="Signature" %}
+```go
+// MultipartForm sends multipart form request with k-v and files.
+//
+// It is recommended obtaining args via AcquireArgs and release it
```
Comment (hi019) [inline api/client.md:324]:
```suggestion
// It is recommended to obtain args via AcquireArgs and release it
```

File: api/client.md [hunk 20]
```diff
@@ -0,0 +1,646 @@
...
+{% code title="Example" %}
+```go
+agent.Boundary("myBoundary")
+    .MultipartForm(nil)
+// ...
+```
+{% endcode %}
+
+#### SendFile(s)
+SendFile(s) read file(s) and appends it(them) to multipart form request.
```
Comment (hi019) [inline api/client.md:361]:
```suggestion
SendFile read a file and appends it to a multipart form request. Sendfiles can be used to append multiple files.
```

**Review by hi019 — CHANGES_REQUESTED** [review]

File: api/client.md (hunk 3 above)
Comment (kiyonlin) [reply api/client.md:29]: > Sounds like Parse takes a URI as a param

Parse parse the total request, users should set URI via agent.Request().

File: api/client.md (hunk 5 above)
Comment (kiyonlin) [reply api/client.md]: Yeah you are right. we can have headers like
```
K1: v1
K1: v2
```

File: api/client.md (hunk 10 above)
Comment (kiyonlin) [reply api/client.md:144]: Yes we can. But if we do so, we have to declare a map struct.

**Pull request merged**